src/openms/source/METADATA/ExperimentalDesign.cpp

```diff
@@ -41,152 +41,578 @@
+      return std::max_element(run_section_.begin(), run_section_.end(),
+                              [](const RunRow& f1, const RunRow& f2)
+                              {
+                                return f1.fraction < f2.fraction;
+                              })->channel;
+    }
+
+    // @return the number of MS files (= fractions * runs)
+    unsigned ExperimentalDesign::getNumberOfMSFiles() const
+    {
+      std::set<std::string> unique_paths;
+      for (auto const & r : run_section_) { unique_paths.insert(r.path); }
+      return unique_paths.size();
+    }
+
+    // @return the number of runs (before fractionation)
+    // Allows to group fraction ids and source files
+    unsigned ExperimentalDesign::getNumberOfPrefractionationRuns() const
+    {
+      if (run_section_.empty()) { return 0; }
+      return std::max_element(run_section_.begin(), run_section_.end(),
+                              [](const RunRow& f1, const RunRow& f2)
+                              {
+                                  return f1.run < f2.run;
+                              })->run;
+    }
+
+    // @return sample index (depends on run and channel)
+    unsigned ExperimentalDesign::getSample(unsigned run, unsigned channel)
+    {
+      return std::find_if(run_section_.begin(), run_section_.end(),
+                          [&run, &channel](const RunRow& r)
+                          {
+                              return r.run == run && r.channel == channel;
+                          })->sample;
+    }
+
+    const ExperimentalDesign::SampleSection& ExperimentalDesign::getSampleSection() const
+    {
+      return sample_section_;
+    }
+
+    void ExperimentalDesign::getFileNames(std::vector< String > &filenames, const bool basename) const
+    {
+      filenames.clear();
+      for (const RunRow &row : run_section_)
+      {
+        const String path = String(row.path);
+        filenames.push_back(basename ? path : File::basename(path));
+      }
+    }
+
+    void ExperimentalDesign::getChannels(std::vector<unsigned> &channels) const
+    {
+      channels.clear();
+      for (const RunRow &row : run_section_)
+      {
+        channels.push_back(row.channel);
+      }
+    }
+
+    void ExperimentalDesign::getFractions(std::vector<unsigned> &fractions) const
+    {
+      fractions.clear();
+      for (const RunRow &row : run_section_)
+      {
+        fractions.push_back(row.fraction);
+      }
+    }
+
+    void ExperimentalDesign::sort_()
+    {
+      std::sort(run_section_.begin(), run_section_.end(),
+                [](const RunRow& a, const RunRow& b)
+                {
+                  return std::tie(a.run, a.fraction, a.channel, a.sample, a.path) <
+                         std::tie(b.run, b.fraction, b.channel, b.sample, b.path);
+                });
+    }
+
+
+    /* Implementations of SampleSection */
+
+    void ExperimentalDesign::SampleSection::getSamples(std::set<unsigned> &samples) const
```

Comment:
better to just return the set of samples?
filling the referenced vector is kind of old-style c++